### ISSUE TITLE
borg2: build_chunkindex_from_repo is slow

### ISSUE THREAD

**Issue by ThomasWaldmann** [borgbackup/borg#8397]
Problem:

That function does a `repository.list()`, listing all the object IDs in the repo to build an in-memory chunkindex.

Because all objects are stored separately into a 2 levels deep dir structure, that are (1+)256+65536 listdir() calls in the worst case. Depending on store speed, connection latency, etc., that can take quite a while.

The in-memory chunkindex is currently not persisted to local cache.

**Comment by ThomasWaldmann** [comment]
Analysis:

There are only a few borg2 commands that remove objects from `data/` in the store:
- borg compact (deletes unused objects)
- borg check (deletes corrupted objects)
- borg debug obj-delete

Notably, these commands do NOT delete objects from `data/`:
- borg delete (just kills the entry in `archives/`)
- borg prune (basically calls borg delete internally)

So, the set of objects in the repo is always increasing until compact/check is run (we can ignore borg debug, these commands are documented to be dangerous and not updating indexes/caches/etc.).

borg create must not assume a chunk is in the repo when it in fact isn't anymore, that would create a corrupt archive.

OTOH, storing a chunk into the repo that already existing in there (but we did not know) is only a performance issue, but otherwise not a problem.

**Comment by ThomasWaldmann** [comment]
Implementation idea:

- uptodate = TBD
- load chunkindex from cache if uptodate else (use build_chunkindex_from_repo + save to disk immediately)  
- update in-memory chunkindex (borg create adds new chunks)
- save chunkindex to disk

Uptodate check and lockless operation (even if multiple borg of same user on same machine use the same repository) needs more thoughts.

**Comment by ThomasWaldmann** [comment]
Another idea:
- borg compact needs to use `repository.list()` anyway (and has an exclusive lock), so it could build the list of objects ids (after it deleted objects due to garbage collection) and store them into the repository (`cache/*` maybe?) and also to local cache.
- format could be a compacted ChunkIndex file
- clients could load that ChunkIndex from local cache or repo.
- borg check --repair would either have to update these caches or invalidate them IF it has deleted objects.

**Comment by SpiritInAShell** [comment]
This is tested on a Hetzner storagebox. Accessing from a 100/50MBit Telekom fibre connection.

If I understand correctly, the process in this issue is logged as following?
```
{
  "type": "log_message",
  "time": 1727224246.2930982,
  "message": "[chan 0] listdir(b'/repo.borg2beta10/data/a3/9e')",
  "levelname": "DEBUG",
  "name": "paramiko.transport.sftp"
}
``` 
As far as I know (asked ChatGPT) the SFTP protocol does not have a "get me all subdirs". This process is painful slow.

The repo size is about 11GiB.

I mounted the remote directory with "sshfs" and did a `time find` and aborted it after 10 minutes (doing another run right now)

I did a `time rsync -avPi [user@server]:/[remote path]` (like a directory listing) and that took about 8 seconds

As every directory read has a significant overhead: what would be if the data chunks were put in single, numerical directories, until a direcory is "full" or "saturated" with a defined number of files?

I see that the chunks are sorted into subdirs related to their filename's first 2 chars and that will be for a reason. **But is it worth the read overhead of sftp protocol? In the end, you create a local client-side index anyway.**

Maybe other access protocols like rclone (which I do not know at all) will not have these limitations and therefor the subdir sctructure will be an advantage.
